src/snowflake/snowpark/_internal/temp_table_auto_cleaner.py

```diff
@@ -0,0 +1,93 @@
+    helps maintain a clean working environment and reduce storage cost for a long-running
+    session.
+    """
+
+    def __init__(self, session: "Session") -> None:
+        self.session = session
+        # this dict maintains key-value pair from Snowpark-generated temp table fully-qualified name
+        # to its reference count for later temp table management
+        # this dict will still be maintained even if the cleaner is stopped (`stop()` is called)
+        self.ref_count_map: Dict[str, int] = defaultdict(int)
+        self.queue: Queue = Queue()
+        self.cleanup_thread: Optional[threading.Thread] = None
+
+    def add(self, table: SnowflakeTable) -> None:
+        self._inc_ref_count(table.name)
+        # the finalizer will be triggered when it gets garbage collected
+        # and this table will be dropped finally
+        _ = weakref.finalize(table, self._dec_ref_count, table.name)
+
+    def _inc_ref_count(self, name: str) -> None:
+        self.ref_count_map[name] += 1
+
+    def _dec_ref_count(self, name: str) -> None:
+        self.ref_count_map[name] -= 1
+        if self.ref_count_map[name] == 0:
+            self.ref_count_map.pop(name)
+            # clean up
+            if self.is_alive():
+                self.queue.put(name)
+        elif self.ref_count_map[name] < 0:
+            logging.debug(
+                f"Unexpected reference count {self.ref_count_map[name]} for table {name}"
+            )
+
+    def process_cleanup(self) -> None:
+        while True:
+            table_name = self.queue.get()
+            if table_name is None:
+                # exit the thread loop
+                break
+            self.drop_table(table_name)
+
+    def drop_table(self, name: str) -> None:
+        common_log_text = f"temp table {name} in session {self.session.session_id}"
+        logging.debug(f"Cleanup Thread: Ready to drop {common_log_text}")
+        try:
+            # TODO SNOW-1556553: Remove this workaround once multi-threading of Snowpark session is supported
+            with self.session._conn._conn.cursor() as cursor:
+                cursor.execute(f"drop table if exists {name}")
+            logging.debug(f"Cleanup Thread: Successfully dropped {common_log_text}")
+        except Exception:
+            logging.warning(
+                f"Cleanup Thread: Failed to drop {common_log_text}"
+            )  # pragma: no cover
+
+    def is_alive(self) -> bool:
+        return self.cleanup_thread is not None and self.cleanup_thread.is_alive()
+
+    def start(self) -> None:
+        if not self.is_alive():
+            self.cleanup_thread = threading.Thread(target=self.process_cleanup)
+            self.cleanup_thread.start()
+
+    def stop(self) -> None:
+        if self.is_alive():
+            # None will break the loop in cleanup thread and force it exit
+            # then we can call thread.join() to stop this thread
+            self.queue.put(None)
```

Comment:
i am not sure if I get this part, were you relying on this code here
```
if table_name is None:
                # exit the thread loop
                break
```
so if the queue by the time when we call stop, still have some items, it will finish those items first until it reaches none? and until it reaches none, there can continue be some items added to the queue if some ref descrease happens? is there a way to force interrupt the thread in python to stop the clean up immediately. if user calls stop, that might means we are doing something wrong, we should stop immediately.

Reply:
First, `stop` and `TempTableAutoCleaner` is not exposed to users so the only way users can "stop" cleaner is to set `session.auto_clean_up_temp_table_enabled = False`. Based on our previous experience of how users are using sql simplifier's parameter, it is more likely to be used:

```
session.parameter = True
... # dataframes code that needs sql simplifier
session.parameter = False
```

so with this context, I feel the cleaner should finish cleanup temp tables that in the queue. But if you prefer to force it exit, we can also do that.